Parser/myreadline.c

```diff
@@ -114,6 +114,8 @@ _PyOS_WindowsConsoleReadline(HANDLE hStdIn)
     wbuf = wbuf_local;
     wbuflen = sizeof(wbuf_local) / sizeof(wbuf_local[0]) - 1;
     while (1) {
+        if (PyOS_InputHook != NULL)
+            (void)(PyOS_InputHook)();
```

Comment:
My only question is whether it's okay to call `PyOS_InputHook` multiple times if we get a partial read below? Or if we should only call it once per `input()` call (in which case, put it outside the `while` loop)

Reply:
I think it needs to be run in the case of a partial read (which I am taking to mean the user has not hit 'return' yet) as while we wait for the user to finish typing `PyOS_InputHook` should be called to let the GUI event loops spin and make the windows seem responsive.

On linux these hooks tend to run forever but exit them selves as soon as there is something to be read on stdin.  I have a vague memory that the hooks work differently on windows where they run for a fixed about of time and expect to be called repeatedly.  If that is the case I am worried this may not actually fix the bug as it looks like `ReadConsoleW` blocks until it has something to read.

Reply:
On a bit more digging, I think my concerns are miss-placed (see the tk EventHook below), but I am pretty sure that this needs to run every time through this loop.


https://github.com/python/cpython/blob/492572715aa0f4ddab51f979f7f56465c762227c/Modules/_tkinter.c#L3301-L3348

Reply:
It certainly fixes the referenced issues, so I approve.

Reply:
@zooba, this change reproduces the legacy implementation, which loops in `PyOS_StdioReadline` until LF is read and calls `PyOS_InputHook` at the start of `my_fgets`. 

Note that with the default console mode and `NULL` `pInputControl` value, `ReadConsoleW` blocks until a CR is typed or written to the input buffer. In most cases this function reads the whole line in one pass since the initial `wbuf_local` is 16K characters. The chance of a single line being longer is negligible. 

For a more responsive interaction, we would need a low-level implementation that calls `ReadConsoleInput`, like what PowerShell uses. This entails developing our own command-line editing interface, i.e. we lose everything that's provided by the console's built-in cooked read (e.g. editing keys such as left/right arrows, backspace, home, end, and escape; input history via up/down arrows and F7; and input aliases). This is actually already implemented by pyreadline using ctypes. An ambitious dev could port this to a C extension that provides built-in readline support in Windows.

Reply:
A less ambitious addition could implement just command completion using the `pInputControl` parameter of `ReadConsoleW`. This is how CMD's tab completion works. It's still a cooked read, so it doesn't do much for responsiveness in terms of `PyOS_InputHook`. At best the hook can be called in each pass through the tab-completion loop. However, the alternative with `ReadConsoleInput` is a major undertaking -- and probably too much to take on. It's not simply providing an interface for a readline library; it's developing the entire library from scratch.

Anyway, here's an overview of how the CMD shell implements tab completion for console input. 

Before the input loop, it saves the initial cursor position from `GetConsoleScreenBufferInfo` in order to support rewriting the line while cycling through possible completions. The `dwCtrlWakeupMask` field of the [`CONSOLE_READCONSOLE_CONTROL`](https://docs.microsoft.com/en-us/windows/console/console-readconsole-control) record is set to return when tab is pressed (i.e. 0x200). After the call returns, the `dwControlKeyState` field is checked for the shift key (0x10), in which case it cycles in reverse.

In the input loop it calls `ReadConsoleW` and then looks for CR and tab in the buffer. CR terminates the read. Otherwise if tab is found, it completes the command in place in the buffer, which for CMD is simply filename completion. Next it updates the screen with the command. For this, it resets the cursor position via `SetConsoleCursorPosition`, overwrites the previous command via `FillConsoleOutputCharacterW`, and writes the new completed command via `WriteConsoleW`. Finally, for the next iteration it makes the console skip what's already read/completed in the buffer by setting the `nInitialChars` field, which I think also integrates with the console history buffer, so it doesn't end up with partial lines.

Reply:
To quantify the difference here, with the Unix readline module, the input hook is called every 0.1 seconds, which is implemented using a timeout in `readline_until_enter_or_signal`. The docs for `PyOS_InputHook` are short on details, so I suppose this is allowed.

As noted above, Tkinter doesn't depend on the input hook being called repeatedly in order to pump events. In Windows the input hook loops so long as `_kbhit` either returns false or fails because stdin isn't console input. We can make this misbehave. Run bash under mintty and force Python to run interactive via `python -i`. Then start Tk via `root = tkinter.Tk()`. The REPL will hang in tkinter's input hook because stdin is a pipe in this scenario.